File: internal/domain/productcollection/productcollection.go
package productcollection

import (
	"dresser/internal/domain/brands"
	"dresser/internal/domain/categoriess"
	"dresser/internal/domain/products"
	"errors"
)

type ProductCollection struct {
	Products []*products.Product
}

func NewProductCollection(products []*products.Product) *ProductCollection {
	return &ProductCollection{
		Products: products,
	}
}

func (pc *ProductCollection) FilterByCategory(category string) *ProductCollection {
	filteredProducts := make([]*products.Product, 0)
	for _, product := range pc.Products {
		if product.Category == category {
			filteredProducts = append(filteredProducts, product)
		}
	}
	return NewProductCollection(filteredProducts)
}

func (pc *ProductCollection) FilterByBrand(brandID brands.ID) *ProductCollection {
	filteredProducts := make([]*products.Product, 0)
	for _, product := range pc.Products {
		if product.Brand == brandID {
			filteredProducts = append(filteredProducts, product)
		}
	}
	return NewProductCollection(filteredProducts)
}

func (pc *ProductCollection) GetLowestPriceProduct() (*products.Product, error) {
	var (
		lowestPriceProduct *products.Product
		lowestPrice        = 100000000
	)
	if len(pc.Products) == 0 {
		return nil, errors.New("no products")
	}

	for _, product := range pc.Products {
		// todo: 화폐 단위 고려
		if product.Price.Amount() < lowestPrice {
			lowestPrice = product.Price.Amount()
			lowestPriceProduct = product
		}
	}
	return lowestPriceProduct, nil
}
func (pc *ProductCollection) GetHighestPriceProduct() (*products.Product, error) {
	var (
		highestPriceProduct *products.Product
		highestPrice        = 0
	)
	if len(pc.Products) == 0 {
		return nil, errors.New("no products")
	}

	for _, product := range pc.Products {
		// todo: 화폐 단위 고려
		if product.Price.Amount() > highestPrice {
			highestPrice = product.Price.Amount()
			highestPriceProduct = product
		}
	}
	return highestPriceProduct, nil
}

func (pc *ProductCollection) GetLowestProductsByCategories() *ProductCollection {
	cs := categoriess.CATEGORIES

	var lowestProducts []*products.Product

	for _, c := range cs {
		p, err := pc.FilterByCategory(string(c)).GetLowestPriceProduct()
		if err != nil {
			continue
		}
		lowestProducts = append(lowestProducts, p)
	}

	return NewProductCollection(lowestProducts)
}

File: java/src/main/java/com/dresser/domain/productcollection/ProductCollection.java
package com.dresser.domain.productcollection;

import com.dresser.domain.brands.BrandId;
import com.dresser.domain.categories.Category;
import com.dresser.domain.products.Product;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class ProductCollection {
    private final String name;
    private final List<Product> products;
    
    private ProductCollection(String name, List<Product> products) {
        if (name == null || name.trim().isEmpty()) {
            throw new IllegalArgumentException("Collection name cannot be null or empty");
        }
        
        this.name = name;
        this.products = new ArrayList<>(products);
    }
    
    public static ProductCollection of(String name, List<Product> products) {
        if (products == null) {
            products = Collections.emptyList();
        }
        return new ProductCollection(name, products);
    }
    
    public String getName() {
        return name;
    }
    
    public List<Product> getProducts() {
        return products;
    }
    
    public int getProductCount() {
        return products.size();
    }
    
    public List<String> getCategories() {
        List<String> categories = new ArrayList<>();
        for (Product product : products) {
            String category = product.getCategory();
            if (!categories.contains(category)) {
                categories.add(category);
            }
        }
        return categories;
    }
    
    public ProductCollection filterByCategory(String category) {
        if (category == null || category.trim().isEmpty()) {
            return this;
        }
        
        List<Product> filtered = products.stream()
                .filter(product -> category.equals(product.getCategory()))
                .collect(Collectors.toList());
        
        return new ProductCollection(this.name + " (Category: " + category + ")", filtered);
    }
    
    public ProductCollection filterByBrand(BrandId brandId) {
        if (brandId == null) {
            return this;
        }
        
        List<Product> filtered = products.stream()
                .filter(product -> brandId.equals(product.getBrandId()))
                .collect(Collectors.toList());
        
        return new ProductCollection(this.name + " (Brand: " + brandId.getValue() + ")", filtered);
    }
    
    public Optional<Product> getLowestPriceProduct() {
        if (products.isEmpty()) {
            return Optional.empty();
        }
        
        return products.stream()
                .filter(product -> !product.isDeleted())
                .min(Comparator.comparing(Product::getPrice));
    }
    
    public Optional<Product> getHighestPriceProduct() {
        if (products.isEmpty()) {
            return Optional.empty();
        }
        
        return products.stream()
                .filter(product -> !product.isDeleted())
                .max(Comparator.comparing(Product::getPrice));
    }
    
    public ProductCollection getLowestProductsByCategories() {
        List<Product> lowestProducts = new ArrayList<>();
        
        // Group products by category
        var productsByCategory = products.stream()
            .collect(Collectors.groupingBy(Product::getCategory));
        
        // Get lowest price product for each category
        for (var entry : productsByCategory.entrySet()) {
            entry.getValue().stream()
                .min(Comparator.comparing(Product::getPrice))
                .ifPresent(lowestProducts::add);
        }
        
        return new ProductCollection(name + " (Lowest by Category)", lowestProducts);
    }
    
    public BigDecimal getTotalPrice() {
        return products.stream()
                .filter(product -> !product.isDeleted())
                .map(Product::getPrice)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }
} 